src/sorted_container_iteration.jl

```diff
@@ -0,0 +1,1167 @@
+## Functions in this file implement many kinds of iterations over
+## SortedDict, SortedMultiDict, and SortedSet.
+## The iteration state is a "semitoken".
+## A "token" is a (container,index_into_container) 2-tuple,
+## while semitokens are the second part, index_into_container.
+
+# From tokens.jl:
+# abstract type AbstractSemiToken end
+
+# struct IntSemiToken <: AbstractSemiToken
+#     address::Int
+# end
+
+const SDMContainer = Union{SortedDict, SortedMultiDict}
+const SortedContainer = Union{SDMContainer, SortedSet}
+const Token = Tuple{SortedContainer, IntSemiToken}
+const SortedDictToken = Tuple{SortedDict, IntSemiToken}
+const SortedMultiDictToken = Tuple{SortedMultiDict, IntSemiToken}
+const SDMToken = Tuple{SDMContainer, IntSemiToken}
+const SortedSetToken = Tuple{SortedSet, IntSemiToken}
+
+(==)(t1::Token, t2::Token) = (t1[1] === t2[1] && t1[2] == t2[2])
+
+
+"""
+    Base.firstindex(m::SortedContainer)
+    startof(m::SortedContainer)
+
+Return the semitoken of
+the first entry of the container `m`, or the past-end semitoken
+if the container is empty.  Time: O(log *n*)
+"""  
+Base.firstindex(m::SortedContainer) = startof(m)
+startof(m::SortedContainer) = IntSemiToken(beginloc(m.bt))
+
+
+"""
+    token_firstindex(m::SortedContainer)
+
+
+Return the token of
+the first entry of the sorted container `m`, or the past-end token
+if the container is empty.  Time: O(log *n*)
+"""  
+token_firstindex(m::SortedContainer) = (m, firstindex(m))
```

Comment:
do we need this to be exported?
It seems like if someone needs it is is a trivial and obvious definition.

Reply:
See other comment about `token_firstindex`.